drivers/gpu/drm/rp1/rp1-vec/rp1_vec.c

```diff
@@ -298,27 +298,43 @@ static enum drm_mode_status rp1vec_mode_valid(struct drm_device *dev,
 					      const struct drm_display_mode *mode)
 {
 	/*
-	 * Check the mode roughly matches one of our standard modes
-	 * (optionally half-height and progressive). Ignore H/V sync
-	 * timings which for interlaced TV are approximate at best.
+	 * Check the mode roughly matches something we can generate.
+	 * The hardware driver is very prescriptive about pixel clocks
+	 * and frame timings, but we'll tolerate some rounding errors.
+	 * It supports a range of frame sizes in each standard (which may be
+	 * useful to fine-tune overscan correction or emulate retro devices).
+	 * Ignore H/V sync timings which for interlaced TV are ill-defined.
 	 */
-	int i, prog;
 
-	prog = !(mode->flags & DRM_MODE_FLAG_INTERLACE);
-
-	for (i = 0; i < ARRAY_SIZE(rp1vec_modes); i++) {
-		const struct drm_display_mode *ref = rp1vec_modes + i;
-
-		if (mode->hdisplay == ref->hdisplay           &&
-		    mode->vdisplay == (ref->vdisplay >> prog) &&
-		    mode->clock + 2 >= ref->clock             &&
-		    mode->clock <= ref->clock + 2             &&
-		    mode->htotal + 2 >= ref->htotal           &&
-		    mode->htotal <= ref->htotal + 2           &&
-		    mode->vtotal + 2 >= (ref->vtotal >> prog) &&
-		    mode->vtotal <= (ref->vtotal >> prog) + 2)
-			return MODE_OK;
+	int prog = !(mode->flags & DRM_MODE_FLAG_INTERLACE);
+	int vtotal_full = mode->vtotal << prog;
+	int vdisplay_full = mode->vdisplay << prog;
+
+	/* Reject very small frames */
+	if (vtotal_full < 256 || mode->hdisplay < 256)
+		return MODE_BAD;
+
+	/* Check lines, frame period (ms) and vertical size limit */
+	if (vtotal_full >= 524 && vtotal_full <= 526 &&
+	    mode->htotal * vtotal_full > 33 * mode->clock &&
+	    mode->htotal * vtotal_full < 34 * mode->clock) {
+		if (vdisplay_full > 486)
+			return MODE_BAD;
+	} else if (vtotal_full >= 624 && vtotal_full <= 626 &&
+		   mode->htotal * vtotal_full > 39 * mode->clock &&
+		   mode->htotal * vtotal_full < 41 * mode->clock) {
+		if (vdisplay_full > 576)
+			return MODE_BAD;
+	} else {
+		return MODE_BAD;
 	}
+
+	/* Check pixel rate (kHz) and horizontal size limit */
+	if (mode->clock == 13500 && mode->hdisplay <= 720)
+		return MODE_OK;
+	if (mode->clock >= 15428 && mode->clock <= 15429 &&
```

Comment:
Very small range.
`if ((mode->clock == 15428 || mode->clock == 15429) ...`
?

Reply:
Our standard VEC clock rate is 108 MHz. Dividing by 8 gives the Rec.601 rate of 13.5MHz, but for emulators I thought it would be nice to offer a faster rate (so that 640 columns will comfortably fit on any CRT). 108/7 is a suitable rate but alas not an integer!

Since the back-end ignores most of the timings anyway, we could make the range broader. So somebody asking for 16MHz (the BBC Micro pixel rate) would get something that worked more or less as expected.